### PR TITLE
Focus inserter toggle when closing the inserter sidebar

### PR DESCRIPTION
**Pull request by jeryj**
## What?

Brings the inserter sidebar closing behavior inline with List View. The inserter sidebar can be closed with escape or the visible close button. Doing this always focuses the inserter toggle button.

## Why?

Consistency in the editor.

## How?

Uses the same method as the List View sidebar.

- Adds an inserterSidebarToggleRef to the store
- Adds this ref to the toggle button
- Removes unnecessary hacks related to closing the sidebar on focus outside that were leftover from needing to close the inserter when focus was outside the inserter

## Testing Instructions

- Open the inserter sidebar
- Escape to close the inserter
- Focus should be on the toggle
- Open the inserter
- Use the close button to close the inserter
- Focus should be on the toggle
- Repeat the above after adding blocks, interacting with the canvas, etc.

### CODE REVIEW

**Comment by jeryj** [comment]
Closing in favor of https://github.com/WordPress/gutenberg/pull/61004

File: packages/editor/src/components/inserter-sidebar/index.js [hunk 1]
```diff
@@ -20,27 +20,48 @@ export default function InserterSidebar( {
...
+				event.preventDefault();
+				closeInserterSidebar();
+			}
+		},
+		[ closeInserterSidebar ]
+	);
+
 	return (
-		<div className="editor-inserter-sidebar">
+		// eslint-disable-next-line jsx-a11y/no-static-element-interactions
```
Comment (scruffian) [inline packages/editor/src/components/inserter-sidebar/index.js:74]: We should address this in a future PR.

**Review by scruffian — APPROVED** [review]
Makes sense to copy the list view implementation for now. Soon I hope we can create a shared component for both.

**Review by mikachan — APPROVED** [review]
I followed the testing instructions and this works well for me, including adding blocks and interacting with the canvas 👏

File: packages/editor/src/components/inserter-sidebar/index.js (hunk 1 above)
Comment (jeryj) [reply packages/editor/src/components/inserter-sidebar/index.js:74]: The fix would be moving this interaction up to where the role for the region on the wrapper already is.

Comment (draganescu) [reply packages/editor/src/components/inserter-sidebar/index.js:74]: > moving this interaction up

Why not do it here? Is it complex?

**Comment by afercia** [comment]
@jeryj thanks for this PR. Can you please clarify fi this aims to solve the regressions introduced in https://github.com/WordPress/gutenberg/pull/61004 ? For clarity and history. Thank you.

File: packages/editor/src/components/inserter-sidebar/index.js [hunk 2]
```diff
@@ -20,45 +17,48 @@ export default function InserterSidebar( {
...
-			const {
-				getBlockRootClientId,
-				__unstableGetEditorMode,
-				getSettings,
-			} = select( blockEditorStore );
+			const { getInserterSidebarToggleRef, getInsertionPoint } = unlock(
+				select( editorStore )
+			);
 			const { get } = select( preferencesStore );
-			const getBlockSectionRootClientId = () => {
```
Comment (ntsekouras) [inline packages/editor/src/components/inserter-sidebar/index.js:32]: Is this removal intentional? It seems it was added just yesterday [here](https://github.com/WordPress/gutenberg/pull/61464).

File: packages/editor/src/components/inserter-sidebar/index.js (hunk 1 above)
Comment (jeryj) [reply packages/editor/src/components/inserter-sidebar/index.js:74]: I think it'll be better to do it when refactoring the list view and inserter to share the same sidebar.

File: packages/editor/src/components/inserter-sidebar/index.js (hunk 2 above)
Comment (jeryj) [reply packages/editor/src/components/inserter-sidebar/index.js:32]: I bet it was accidentally removed during a rebase. Nice catch!

File: packages/edit-widgets/src/components/header/document-tools/index.js [hunk 3]
```diff
@@ -4,90 +4,70 @@
...
+
+	const {
+		isInserterOpen,
+		isListViewOpen,
+		inserterSidebarToggleRef,
+		listViewToggleRef,
+	} = useSelect( ( select ) => {
+		const {
+			isInserterOpened,
+			getInserterSidebarToggleRef,
```
Comment (ntsekouras) [inline packages/edit-widgets/src/components/header/document-tools/index.js:31]: e2e failures are valid and are from `getInserterSidebarToggleRef`, which is not present in edit-widgets store.

Comment (jeryj) [reply packages/edit-widgets/src/components/header/document-tools/index.js:31]: Fixed by https://github.com/WordPress/gutenberg/pull/61467/commits/5237640985c9e53e31e738ed975066c3a09f8462

**Comment by alexstine** [comment]
Could someone please link me to the discussion where the block inserter and list view will share the same sidebar? I want the background info on why that change is being considered. It will create a jumbled mess for screen reader users unless we do this right and really consider how we're going to drive separation.

File: packages/editor/src/store/private-selectors.js [hunk 4]
```diff
@@ -75,6 +75,9 @@ export const getInsertionPoint = createRegistrySelector( ( select ) =>
 export function getListViewToggleRef( state ) {
 	return state.listViewToggleRef;
 }
+export function getInserterSidebarToggleRef( state ) {
+	return state.inserterSidebarToggleRef;
```
Comment (kevin940726) [inline packages/editor/src/store/private-selectors.js:79]: I know this is an existing pattern, but I just want to share my concerns about this approach again. I still think storing non-serializable data in the store is a [bad practice](https://redux.js.org/style-guide/#do-not-put-non-serializable-values-in-state-or-actions).

First of all, it breaks the Redux Devtools; I think I have had to disable the extension ever since we merged this approach the first time. Even if we special case this by making it serializable in the [extension settings](https://github.com/reduxjs/redux-devtools/blob/main/extension/docs/API/Arguments.md#serialize), we can't make sure that new features of the extensions will continue support this pattern.

Second, I think using a context provider is a much simpler and intuitive solution. It might mean an extra wrapper and private settings in the `editor` component, but it also matches more closely to what React expects. We can also simply group the same patterns over and over again and create an abstraction around it, so it feels less "hacky."

This is not a blocker and I just want to re-share this for visibility 😆.

File: packages/editor/src/store/reducer.js [hunk 5]
```diff
@@ -360,6 +360,17 @@ export function listViewToggleRef( state = { current: null } ) {
...
 }
 
+/**
+ * This reducer does nothing aside initializing a ref to the inserter sidebar toggle.
+ * We will have a unique ref per "editor" instance.
+ *
+ * @param {Object} state
+ * @return {Object} Reference to the inserter sidebar toggle button.
+ */
+export function inserterSidebarToggleRef( state = { current: null } ) {
```
Comment (kevin940726) [inline packages/editor/src/store/reducer.js:370]: Not sure if it matters, but every time this function is called, the state will be a different instance of ref. Probably better this way:

```suggestion
const INITIAL_REF = createRef( null );
export function inserterSidebarToggleRef( state = INITIAL_REF ) {
```

**Comment by jeryj** [comment]
@alexstine - I'm not sure where discussion was. The idea is to refactor them to share the same wrapping component instead of three or four that are almost identical but using different components with repeated work. The behavior of List View should not be changed.

**Comment by alexstine** [comment]
@jeryj So is this combining code or are we going to have all these components imported into one sidebar? Combining code makes sense, combining UI does not.

**Comment by afercia** [comment]
> I'm not sure where discussion was. 

I'd agree with @alexstine as I'd like to have more background info and context about this change. I'm not sure pushing important changes without an open, broad, discussion is any ideal in a collaborative open source project. Alas, this is only one of the multitude of changes that are just pushed too fast and, in my opinion, too lightly in this project.

That said, I do realize this PR is not the right place for this kind of discussion.

Anyways, I think there's the need for a follow-up to address a couple long standing issues:
- Both the Blok Inserter and the List View render within a labeled `role="region"`, which are ARIA landmarks. These ARIA landmarks appear and disappear dynamically, which defeats the purpose of using ARIA landmarks in the first place. Instead, ARIA landmarks should always be rendered. See for example how the Settings panel and the Publish panel work.
- Refine how the 'Always open list view' should work in combination with the block inserter, see https://github.com/WordPress/gutenberg/issues/61498

**Comment by jeryj** [comment]
This PR doesn't do anything to combine the sidebar code, it's fixing an accessibility issue with focus management using the same method that the list view uses. I think they would be easier to maintain if they shared the same code. I agree it needs to be discussed before doing it. Here's an issue to discuss it: https://github.com/WordPress/gutenberg/issues/61549

**Comment by alexstine** [comment]
Thanks @jeryj . Just wanted to make sure it didn't get lost in the shuffle.
